### DISCUSSION TITLE
Type consistency when using setValue

### DISCUSSION THREAD

**Discussion by roypearce** [react-hook-form/react-hook-form#8086] [Q&A]
[CodeSandbox example](https://codesandbox.io/s/react-hook-form-get-started-forked-fhzsp?file=/src/index.tsx)

I'm doing a trial upgrade from v6 to v7 and noticed that when a value is set with `setValue` the validation, submit handler, and `getValues` preserve the type of that value when it is returned. However if it is set with `value` or typed into the field, then it is always returned as a string. 

Is this something that is known and going to be addressed in v8 or is it a bug in v7? Even with a fully Typescript based project this is slipping through and causing unexpected issues because the project it's used in was written with the expectation that all the data coming out of RHF uncontrolled form elements would be strings, and that held true in v6.

Is there a best practice for how to deal with this? Other than converting all values to string before using `setValue`.

**Comment by Moshyfawn** [comment]
Native HTML input elements always return their value of type `string`. If using `register`, RHF embraces the native Web behavior. You can set register options like
```js
{...register("field", {
  valueAsNumber: true,
  // or
  valueAsDate: true
})}
```
to set the interpreted field's value by using the native `event.target.valueAsDate` and `event.target.valueAsNumber` API.

Additionally you can provide your own custom parser callback
```js
{...register("field", {
  setValueAs: booleanValue => !!booleanValue
})}
```

You can learn more [at RHF docs in the options section](https://react-hook-form.com/api/useform/register/#options)

**Reply by roypearce** [reply]
> Native HTML input elements always return their value of type string. If using register, RHF embraces the native Web behavior. You can set register options like

Yes @Moshyfawn  that is my expectation, that is why when using `setValue`  to set values, and then submitting the form or using another method like `getValues` to retrieve the values, their types are unexpectedly not string in v7. They were strings in v6.

Check out the sandbox, open the console and click the `Get Values...` button, this is expected. Then click the `Set age to 22....` button and then the `Get Values...` button and not the types in the console. That is not expected to me. My question is, is that expected or a bug?

**Reply by Moshyfawn** [reply]
I can't reproduce the V6 behavior you're describing. See [this CSB](https://codesandbox.io/s/react-hook-form-js-tpzo4?file=/src/App.js).

**Reply by Moshyfawn** [reply]
Ah, sry, you're talking about setting **the registered** field value via `setValue`, correct?

**Reply by roypearce** [reply]
> Ah, sry, you're talking about setting the registered field value via setValue, correct?

Correct, [here's a v6 repro](https://codesandbox.io/s/react-hook-form-get-started-forked-51juiz?file=/src/index.tsx) that shows they are always strings under the same scenario.
